src/typeahead/typeahead.ts

```diff
@@ -0,0 +1,132 @@
+import {
+  Directive,
+  OnInit,
+  Input,
+  ComponentRef,
+  ComponentFactoryResolver,
+  ViewContainerRef,
+  Injector,
+  Renderer,
+  ElementRef,
+  TemplateRef
+} from '@angular/core';
+import {Observable} from 'rxjs/Rx';
+import 'rxjs/add/operator/filter';
+import 'rxjs/add/operator/let';
+import {Positioning} from '../util/positioning';
+import {NgbTypeaheadWindow, ResultTplCtx} from './typeahead-window';
+import {PopupService} from '../util/popup';
+import {NgModel} from '@angular/common';
+import {isString} from '../util/util';
+import {NgbTypeaheadValueAccessor} from './typeahead-valueaccessor';
+
+/**
+ * NgbTypeahead directive provides a simple way of creating powerful typeaheads from any text input
+ */
+@Directive({
+  selector: 'input[ngbTypeahead][ngModel]',
+  host: {
+    '[class.open]': '_windowRef !== null',
+    '(document:click)': '_closePopup()',
+    '(keydown.arrowDown)': '_onKeyDown($event)',
+    '(keydown.arrowUp)': '_onKeyUp($event)',
+    '(keydown.enter)': '_onKeyEnter()',
+    '(keydown.esc)': '_closePopup()'
+  }
+})
+export class NgbTypeahead implements OnInit {
+  private _popupService: PopupService<NgbTypeaheadWindow>;
+  private _positioning = new Positioning();
+  private _windowRef: ComponentRef<NgbTypeaheadWindow>;
+
+  /**
+   * A function to transform the provided observable text into the array of results
+   */
+  @Input() ngbTypeahead: (text: Observable<string>) => Observable<any[]>;
+
+  /**
+   * A function to format a given result before display. This function should return a formatted string without any
+   * HTML markup.
+   */
+  @Input() resultFormatter: (value) => string;
+
+  /**
+   * A template to display a matching result.
+   */
+  @Input() resultTemplate: TemplateRef<ResultTplCtx>;
+
+  constructor(
+      private _model: NgModel, private _elementRef: ElementRef, private _viewContainerRef: ViewContainerRef,
+      private _renderer: Renderer, private _injector: Injector, componentFactoryResolver: ComponentFactoryResolver) {
+    this._popupService = new PopupService<NgbTypeaheadWindow>(
+        NgbTypeaheadWindow, _injector, _viewContainerRef, _renderer, componentFactoryResolver);
+  }
+
+  ngOnInit() {
+    this._model.control.valueChanges.filter((value) => isString(value)).let (this.ngbTypeahead).subscribe((results) => {
+
+      if (!results || results.length === 0) {
+        this._closePopup();
+      } else {
+        this._openPopup();
+        this._windowRef.instance.results = results;
+        this._windowRef.instance.term = this._model.value;
+        if (this.resultFormatter) {
+          this._windowRef.instance.formatter = this.resultFormatter;
+        }
+        if (this.resultTemplate) {
+          this._windowRef.instance.resultTemplate = this.resultTemplate;
+        }
+      }
+    });
+  }
+
+  private _onKeyEnter() {
+    if (this._windowRef) {
+      const result = this._windowRef.instance.getActive();
+      this._selectResult(result);
```

Comment:
Does it make sense to call `event.preventDefault();` in here (so for example a form doesn't get submitted?)